src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -147,15 +147,12 @@ public async Task<HttpResponseMessage> SendAsync(CancellationToken cancellationT
                     // Ideally, headers will be sent out in a gathered write inside of SendContentAsync().
                     // If we don't have content, or we are doing Expect 100 Continue, then we can't rely on
                     // this and must send our headers immediately.
+                    await FlushSendBufferAsync(requestCancellationSource.Token).ConfigureAwait(false);
 
-                    await _stream.WriteAsync(_sendBuffer.ActiveMemory, endStream: _expect100ContinueCompletionSource == null, requestCancellationSource.Token).ConfigureAwait(false);
-                    _sendBuffer.Discard(_sendBuffer.ActiveLength);
-
-                    if (_expect100ContinueCompletionSource != null)
+                    // End the stream writing if there's no content to send.
+                    if (_request.Content == null)
                     {
-                        // Flush to ensure we get a response.
-                        // TODO: MsQuic may not need any flushing.
-                        await _stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+                        _stream.Shutdown();
```

Comment:
@ManickaP  It should not be a separate operation -- if there's no content, we need to send FIN flag together with headers as discussed here https://github.com/dotnet/runtime/issues/55347
(current change effectively reverted https://github.com/dotnet/runtime/pull/56221)

Reply:
Ha, and I was like, well it doesn't matter, shutdown will send the FIN :smile: 
I'll put up a pr with a fix in a moment. I'll also add a comment to prevent similar reverts in the future. Good catch BTW, thanks!